File: src/models/Evaluation.ts
// ============================================
// Evaluation Model
// ============================================

export interface EvaluationCriteria {
  pronunciation: number; // 0-100
  grammar: number; // 0-100
  vocabulary: number; // 0-100
  fluency: number; // 0-100
  participation: number; // 0-100
  comprehension: number; // 0-100
}

export interface EvaluationModel {
  id: string;
  sessionId: string;
  studentId: string;
  professorId: string;
  criteria: EvaluationCriteria;
  overallScore: number;
  feedback: string;
  strengths: string[];
  areasToImprove: string[];
  createdAt: string;
  updatedAt?: string;
}

export interface CreateEvaluationDTO {
  sessionId: string;
  studentId: string;
  criteria: EvaluationCriteria;
  feedback: string;
  strengths?: string[];
  areasToImprove?: string[];
}

export interface UpdateEvaluationDTO {
  criteria?: Partial<EvaluationCriteria>;
  feedback?: string;
  strengths?: string[];
  areasToImprove?: string[];
}

export interface EvaluationFilters {
  sessionId?: string;
  studentId?: string;
  professorId?: string;
  fromDate?: string;
  toDate?: string;
  minScore?: number;
  page?: number;
  limit?: number;
}

// Calculate overall score from criteria
export const calculateOverallScore = (criteria: EvaluationCriteria): number => {
  const weights = {
    pronunciation: 0.2,
    grammar: 0.2,
    vocabulary: 0.15,
    fluency: 0.2,
    participation: 0.1,
    comprehension: 0.15,
  };

  return Math.round(
    criteria.pronunciation * weights.pronunciation +
    criteria.grammar * weights.grammar +
    criteria.vocabulary * weights.vocabulary +
    criteria.fluency * weights.fluency +
    criteria.participation * weights.participation +
    criteria.comprehension * weights.comprehension
  );
};

File: backend-spring/src/main/java/com/lingua/hub/entity/Evaluation.java
package com.lingua.hub.entity;

import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Type;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "evaluations", uniqueConstraints = {
    @UniqueConstraint(columnNames = {"session_id", "student_id"})
})
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Evaluation {
    
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "session_id", nullable = false)
    private Room session;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "professor_id")
    private Professor professor;
    
    @Column
    private Integer pronunciation;
    
    @Column
    private Integer grammar;
    
    @Column
    private Integer vocabulary;
    
    @Column
    private Integer fluency;
    
    @Column
    private Integer participation;
    
    @Column
    private Integer comprehension;
    
    @Column(name = "overall_score")
    private Integer overallScore;
    
    @Column(columnDefinition = "TEXT")
    private String feedback;
    
    @Type(JsonType.class)
    @Column(columnDefinition = "jsonb")
    private List<String> strengths;
    
    @Type(JsonType.class)
    @Column(name = "areas_to_improve", columnDefinition = "jsonb")
    private List<String> areasToImprove;
    
    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
